### PR TITLE
fix: do not inject GIT_USERNAME and GIT_PASSWORD into git clone URL

### PR DESCRIPTION
**Pull request by johnstcn**
Fixes https://github.com/coder/envbuilder/issues/126

We had been writing the `GIT_USERNAME` and `GIT_PASSWORD` into the git auth URL as well as setting basic HTTP auth credentials.
This causes `GIT_USERNAME` and `GIT_PASSWORD` to be present in the `.git/config` of the clond repo, which is not desirable. 
This PR removes that behaviour and modifies existing unit tests to assert that the git clone URL is not modified from what the user passes in.

The previous behaviour can be preserved by explicitly setting basic auth in the git clone URL, if required.

### CODE REVIEW

File: integration/integration_test.go [hunk 1]
```diff
@@ -756,27 +780,12 @@ type gitServerOptions struct {
...
 	t.Helper()
 	if opts.authMW == nil {
-		opts.authMW = checkBasicAuth(opts.username, opts.password)
+		opts.authMW = gittest.BasicAuthMW(opts.username, opts.password)
 	}
 	srv := httptest.NewServer(opts.authMW(createGitHandler(t, opts)))
 	return srv.URL
 }
 
-func checkBasicAuth(username, password string) func(http.Handler) http.Handler {
```
Comment (johnstcn) [inline integration/integration_test.go:765]: review: moved to `gittest`

File: git_test.go [hunk 2]
```diff
@@ -4,72 +4,198 @@ import (
...
+	githttp "github.com/go-git/go-git/v5/plumbing/transport/http"
 	"github.com/stretchr/testify/require"
 )
 
 func TestCloneRepo(t *testing.T) {
 	t.Parallel()
 
-	t.Run("Clones", func(t *testing.T) {
-		t.Parallel()
+	for _, tc := range []struct {
```
Comment (johnstcn) [inline git_test.go:25]: review: trying to cover all cases outlined here: 

https://github.com/coder/envbuilder?tab=readme-ov-file#git-authentication

File: envbuilder.go [hunk 3]
```diff
@@ -361,12 +361,12 @@ func Run(ctx context.Context, options Options) error {
...
-				return fmt.Errorf("parse git url: %w", err)
-			}
-			gitURL.User = url.UserPassword(options.GitUsername, options.GitPassword)
-			options.GitURL = gitURL.String()
+			// Previously, we had been placing credentials in the URL
+			// as well as setting githttp.BasicAuth.
+			// This was removed as it would leak the credentials used
+			// to clone the repo into the resulting workspace.
+			// Users may still hard-code credentials directly into the
+			// git URL themselves, if required.
```
Comment (johnstcn) [inline envbuilder.go:369]: review: alternatively, we can keep this and munge the git config afterwards.

File: integration/integration_test.go [hunk 4]
```diff
@@ -71,13 +71,37 @@ func TestSucceedsGitAuth(t *testing.T) {
...
+	u, err := url.Parse(gitURL)
+	require.NoError(t, err)
+	u.User = url.UserPassword("kyle", "testing")
+	ctr, err := runEnvbuilder(t, options{env: []string{
+		"GIT_URL=" + u.String(),
+		"DOCKERFILE_PATH=Dockerfile",
+	}})
+	require.NoError(t, err)
+	gitConfig := execContainer(t, ctr, "cat /workspaces/.git/config")
+	require.Contains(t, gitConfig, u.String())
```
Comment (johnstcn) [inline integration/integration_test.go:104]: review: Check here is to ensure that we don't modify git clone URL

File: gittest/gittest.go [hunk 5]
```diff
@@ -118,3 +118,18 @@ func WriteFile(t *testing.T, fs billy.Filesystem, path, content string) {
 	err = file.Close()
 	require.NoError(t, err)
 }
+
+func BasicAuthMW(username, password string) func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			if username != "" && password != "" {
```
Comment (dannykopping) [inline gittest/gittest.go:125]: Wouldn't this mean that if you pass either `username` or `password` then auth will be bypassed?

```suggestion
			if username != "" || password != "" {
```

File: integration/integration_test.go (hunk 4 above)
Comment (dannykopping) [reply integration/integration_test.go:104]: I'd suggest moving this to a comment.

File: git_test.go [hunk 6]
```diff
@@ -4,72 +4,198 @@ import (
...
+					return
+				}
+				require.NoError(t, err)
+				require.True(t, cloned)
+
+				readme := mustRead(t, clientFS, "/workspace/README.md")
+				require.Equal(t, "Hello, world!", readme)
+				gitConfig := mustRead(t, clientFS, "/workspace/.git/config")
+				// We do not modify the git URL that folks pass in.
+				require.Contains(t, gitConfig, authURL.String())
```
Comment (dannykopping) [inline git_test.go:163]: This may be a bit flimsy; you could harden it with:

```suggestion
				require.Regexp(t, fmt.Sprintf("^\s*url = %s$", regexp.QuoteMeta(authURL.String()), gitConfig)
```

**NOTE**: I have not tested the above code

File: envbuilder.go (hunk 3 above)
Comment (mtojek) [reply envbuilder.go:369]: Maybe rephrase it a bit: "Do not keep credentials..." or add an extra panic when it happens?

File: integration/integration_test.go (hunk 1 above)
Comment (mtojek) [reply integration/integration_test.go:765]: ~~is it intended to be used anywhere else in envbuilder? I'm considering moving it to `integration/gittest.go`~~

Ok, I found the place. Speaking of consistency with coder/coder that should land in `testutil`.

File: git_test.go [hunk 7]
```diff
@@ -4,72 +4,198 @@ import (
...
+func mustRead(t *testing.T, fs billy.Filesystem, path string) string {
+	t.Helper()
+	f, err := fs.OpenFile(path, os.O_RDONLY, 0644)
+	require.NoError(t, err)
+	content, err := io.ReadAll(f)
+	require.NoError(t, err)
+	return string(content)
+}
+
+func setupGit(t *testing.T, user, pass string) (url string) {
```
Comment (mtojek) [inline git_test.go:178]: What about moving this to `gittest`?

Comment (johnstcn) [reply git_test.go:178]: I'll open a a separate PR to refactor this 👍

File: integration/integration_test.go (hunk 1 above)
Comment (johnstcn) [reply integration/integration_test.go:765]: Will do in a follow-up PR 👍

File: git_test.go [hunk 8]
```diff
@@ -4,72 +4,198 @@ import (
...
+			password:    "",
+			expectClone: true,
+		},
+		{
+			name:        "password only",
+			srvUsername: "",
+			srvPassword: "password",
+			username:    "",
+			password:    "password",
+			expectClone: true,
```
[the remark this answers is not in the document]
Comment (johnstcn) [reply git_test.go:86]: Hm, good point. Once you get into these sort of weeds your testing the tests. Will remove them.

[the remark this answers is not in the document]
Comment (johnstcn) [reply git_test.go:86]: We'll keep user-only requests as that's sort of how the GitHub/Azure DevOps authentiation works.

File: envbuilder.go (hunk 3 above)
Comment (johnstcn) [reply envbuilder.go:369]: I'll just link to the PR with a short note about previous behaviour.

File: gittest/gittest.go (hunk 5 above)
Comment (johnstcn) [reply gittest/gittest.go:125]: whoops!

File: git_test.go (hunk 6 above)
Comment (johnstcn) [reply git_test.go]: Beefed up the regexp a bit

**Review by mtojek — APPROVED** [review]
Thanks for addressing the comments so quickly!
